tests/fault_tolerance/etcd_ha/utils.py

```diff
@@ -156,123 +161,217 @@ def __init__(
+                    logger.info(
+                        f"Removing member with ID {old_member_id} (hex: {hex_member_id})"
+                    )
+                    remove_result = subprocess.run(
+                        ["etcdctl", "member", "remove", hex_member_id],
+                        env=etcdctl_env,
+                        capture_output=True,
+                        text=True,
+                        timeout=5,
+                    )
+                    if remove_result.returncode != 0:
+                        raise RuntimeError(
+                            f"Failed to remove old member: {remove_result.stderr}"
+                        )
+                    logger.info(f"Successfully removed old member {name}")
+        except Exception as e:
+            raise RuntimeError(f"Error during member removal: {e}")
 
-        if leader_idx is None:
-            logger.warning("No leader found to terminate")
-            return None
+        # Add the new member to the cluster
+        logger.info(f"Adding new member {name} to cluster with peer URL {peer_url}")
+        try:
+            add_result = subprocess.run(
+                ["etcdctl", "member", "add", name, f"--peer-urls={peer_url}"],
+                env=etcdctl_env,
+                capture_output=True,
+                text=True,
+                timeout=5,
+            )
+            if add_result.returncode != 0:
+                raise RuntimeError(f"Failed to add new member: {add_result.stderr}")
+            logger.info(f"Successfully added new member {name}")
+        except Exception as e:
+            raise RuntimeError(f"Error adding new member: {e}")
+
+    def start(self):
+        """Start ETCD cluster with configured number of replicas"""
+        logger.info(f"Starting {self.num_replicas}-node ETCD cluster")
 
-        logger.info(f"Terminating current leader: etcd-{leader_idx}")
-        replica = self.replicas[leader_idx]
+        # Start each replica
+        for i in range(self.num_replicas):
+            replica = self._start_replica(i, cluster_state="new")
+            self.replicas.append(replica)
 
-        if replica:
-            replica.__exit__(None, None, None)
-            self.replicas[leader_idx] = None
-            logger.info(f"Leader etcd-{leader_idx} has been terminated")
+        logger.info(f"All {self.num_replicas} ETCD replicas started successfully")
 
-        return leader_idx
+        # Wait for cluster to stabilize
+        self._wait_for_healthy_cluster()
 
     def get_client_endpoints(self) -> List[str]:
         """Get list of active client endpoints"""
         endpoints = []
         for i, replica in enumerate(self.replicas):
             if replica:  # Only include active replicas
-                client_port = self.base_client_port + i
-                endpoints.append(f"http://localhost:{client_port}")
+                client_port = self.base_port + (2 * i)
```

Comment:
Just curious, why the change from `+ i` to `+ 2i`?

Reply:
I used to put two base ports, one for client and the other one for peer. I refactored the two base port address into one, which I hope this will help with @keivenchang work on randomizing port numbers, that we only need a continuous block of ports starting from the base port that can be randomized. For example, if we have 3 etcd replicas, we need a block of 2379, 2380, ..., 2384, which
* etcd-0 uses 2379 as client and 2380 as peer
* etcd-1 uses 2381 as client and 2382 as peer
* etcd-2 uses 2383 as client and 2384 as peer

The `2i` basically jumps over the peer port of the previous etcd replica(s) given the base port.